### ISSUE TITLE
Why no async methods?

### ISSUE THREAD

**Issue by pvanroos** [litedb-org/LiteDB#1300]
Why not implement async Task in the read and writes?  e.g. InsertAsync() or FindByIdAsync()

BTW, love the package!  👍

**Comment by andrewvk** [comment]
https://github.com/rsdn/CodeJam/blob/master/CodeJam.Main/Threading/AsyncLock.cs
You may copy source code to your project.

**Comment by clement128** [comment]
hi @mbdavid  any plan to support async methods?

BTW thanks for the great works

**Comment by mbdavid** [comment]
Hi @clement128,

Was not possible to implement in v5 because was pretty complex at the moment. But now, with v5 released It's on plans no v.5.5 (next "major" upgrade)

**Comment by ansarizafar** [comment]
Here is an Aysnc wrapper for liteDb https://github.com/mlockett42/litedb-async

**Comment by LifshesOfer** [comment]
Hi @mbdavid,

Sorry about necro'ing an old thread, but is there any word about async methods in LiteDB?

**Comment by mbdavid** [comment]
Hi @LifshesOfer, it's an old thread but still important 😄 !! I'm writing a complete new engine for LiteDB, using only async methods and simple datafile structure. It's in this "temp" repository: https://github.com/mbdavid/LiteDB-vNext

**Comment by JamesW940516** [comment]
Hello @mbdavid 
You are doing really great work
Cheers!!

**Comment by mbdavid** [comment]
Hi @JamesW940516, my plans are release first beta version next month. Maybe 60 days to release final version (if not found any tricky problem 😄)

**Comment by JamesW940516** [comment]
> Hi @JamesW940516, my plans are release first beta version next month. Maybe 60 days to release final version (if not found any tricky problem 😄)

Thank you

**Comment by JamesW940516** [comment]
Hi @mbdavid 
Hope you are doing great

I am using litedb.5.0.17 in my unity game.
There are considerably a lot of read/write to the database file.
Unfortunately I got this error.

![image](https://github.com/mbdavid/LiteDB/assets/133299884/7986a0f1-ff91-477c-9b0e-3e1676a98847)
![image](https://github.com/mbdavid/LiteDB/assets/133299884/74a5d372-f160-49fc-bbce-b7bc9572ae66)

`
IOException: Sharing violation on path E:\_integration3_user_collection.db
System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean anonymous, System.IO.FileOptions options) (at <b89873cb176e44a995a4781c7487d410>:0)
System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.IO.FileOptions options) (at <b89873cb176e44a995a4781c7487d410>:0)
(wrapper remoting-invoke-with-check) System.IO.FileStream..ctor(string,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare,int,System.IO.FileOptions)
LiteDB.Engine.FileStreamFactory.GetStream (System.Boolean canWrite, System.Boolean sequencial) (at <f6a00fdda8d849629324c7172d2221b6>:0)
LiteDB.Engine.StreamPool+<>c__DisplayClass3_0.<.ctor>b__0 () (at <f6a00fdda8d849629324c7172d2221b6>:0)
System.Lazy`1[T].ViaFactory (System.Threading.LazyThreadSafetyMode mode) (at <b89873cb176e44a995a4781c7487d410>:0)
System.Lazy`1[T].ExecutionAndPublication (System.LazyHelper executionAndPublication, System.Boolean useDefaultConstructor) (at <b89873cb176e44a995a4781c7487d410>:0)
System.Lazy`1[T].CreateValue () (at <b89873cb176e44a995a4781c7487d410>:0)
System.Lazy`1[T].get_Value () (at <b89873cb176e44a995a4781c7487d410>:0)
LiteDB.Engine.StreamPool.get_Writer () (at <f6a00fdda8d849629324c7172d2221b6>:0)
LiteDB.Engine.DiskService..ctor (LiteDB.Engine.EngineSettings settings, System.Int32[] memorySegmentSizes) (at <f6a00fdda8d849629324c7172d2221b6>:0)
LiteDB.Engine.LiteEngine..ctor (LiteDB.Engine.EngineSettings settings) (at <f6a00fdda8d849629324c7172d2221b6>:0)
LiteDB.ConnectionString.CreateEngine () (at <f6a00fdda8d849629324c7172d2221b6>:0)
LiteDB.LiteDatabase..ctor (LiteDB.ConnectionString connectionString, LiteDB.BsonMapper mapper) (at <f6a00fdda8d849629324c7172d2221b6>:0)
LiteDB.LiteDatabase..ctor (System.String connectionString, LiteDB.BsonMapper mapper) (at <f6a00fdda8d849629324c7172d2221b6>:0)
ServiceManager.FetchBalance (System.UInt64 pUserOwnerClientId) (at Assets/Scripts/Server/ServiceManager.cs:516)
RequestManager.FetchBalance () (at Assets/Scripts/Server/RequestManager.cs:186)
System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) (at <b89873cb176e44a995a4781c7487d410>:0)
UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () (at <e97d84204f8d4aef92b538c5bab948f1>:0)
UnityEngine.UnitySynchronizationContext.Exec () (at <e97d84204f8d4aef92b538c5bab948f1>:0)
UnityEngine.UnitySynchronizationContext.ExecuteTasks () (at <e97d84204f8d4aef92b538c5bab948f1>:0)
`

I know that the reason why I had this error.
This is because I tried to access(read/write) at the same time.
In other words, I tried to open a file which is opened already.
I also know this issue will be resolved after the async litedb is released.

I found my temp solution.
I open the database file at the beginning time of my game and close it at the closing time of my game.
Not open/close the database file for each read/write.
I do not any outside access to the database file when the game is running.

What I am asking if this is possible? Could be a temp solution for now?
And how to keep opening the database file and how to close when I need to close.

Thank you.

**Comment by tinmac** [comment]
@mbdavid Just a quick thank you - appreciate the project & the major async vNext rewrite

**Comment by JamesW940516** [comment]
vNext is release?

**Comment by mbdavid** [comment]
Hi guys, v6 are not ready yet. I merge my code to `dev` branch in this repo. I'm almost with database engine...  My goal is release in this year the final version

**Comment by JamesW940516** [comment]
Thank you @mbdavid

**Comment by JKamsker** [comment]
Consolidated into #2781, which combines this report with the related ones and tracks the fix. Closing this as a duplicate — please follow #2781, and comment there if your case is different.
